File: Parallel-Java/src/parallel/linq101/mylinq/FirstMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Order;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.*;

public class FirstMethods {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public int numbersFirst() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).findFirst().getAsInt();
        return firstNum;
    }

    public int bigNumbersFirst() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).filter(n -> n > 5).findFirst().getAsInt();
        return firstNum;
    }

    public int oddNumbersFirst() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).filter(n -> n % 2 == 1).findFirst().getAsInt();
        return firstNum;
    }

    public double doubleFirst() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).findFirst().getAsDouble();
        return firstNum;
    }

    public double bigDoubleNumbersFirst() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).filter(d -> d > 5).findFirst().getAsDouble();
        return firstNum;
    }

    public String stringFirstStartWithO() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startWithO = Arrays.stream(strings).filter(s -> s.charAt(0) == 'o').findFirst().get();
        return startWithO;
    }

    public String stringFirstStartWith(char c) {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startWithChar = Arrays.stream(strings).filter(s -> s.charAt(0) == c).findFirst().get();
        return startWithChar;
    }

    public String stringFirstLong() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startsWithO = Arrays.stream(strings).filter(s -> s.length() > 4).findFirst().get();
        return startsWithO;
    }

    public int projectedBigNumbersFirst() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).map(i -> (int) Math.pow(2, i)).filter(n -> n > 100).findFirst().getAsInt();
        return firstNum;
    }

    public double projectedDoubleFirst() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).map(i -> Math.log(i)).filter(n -> n > 3).findFirst().getAsDouble();
        return firstNum;
    }

    public int projectedStringFirst() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        int firstStringLength = Arrays.stream(strings).mapToInt(s -> s.length()).filter(i -> i % 2 == 0).findFirst().getAsInt();
        return firstStringLength;
    }

    public Product productFirst() {
        List<Product> products = getProductList();
        Product product12 = products.stream().filter(p -> p.getProductId() == 12).findFirst().get();
        return product12;
    }

    public List<Pair<String, Product>> productEachCategoryFirst() {
        List<Product> products = getProductList();
        List<Pair<String, Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().findFirst().get())).toList();
        return categories;
    }

    public Map<String, Product> productEachCategoryFirstDict() {
        List<Product> products = getProductList();
        Map<String, Product> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().findFirst().get()));
        return categories;
    }

    public List<Pair<String, Order>> customerEachRegionFirstOrder() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Order>> ordersFirst = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().flatMap(c -> c.getOrders().stream()).findFirst().get())).toList();
        return ordersFirst;
    }

    public Map<String, Order> customerEachRegionFirstOrderDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Order> ordersFirst = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().flatMap(c -> c.getOrders().stream()).findFirst().get()));
        return ordersFirst;
    }

    public List<Pair<String, List<Product>>> productEachCategoryFirstUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(p ->
                        p.getUnitPrice() == g.getValue().stream().findFirst().get().getUnitPrice()).toList())).toList();
        return categories;
    }

    public Map<String, List<Product>> productEachCategoryFirstUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().filter(p ->
                                p.getUnitPrice() == g.getValue().stream().findFirst().get().getUnitPrice()).toList()));
        return categories;
    }

    public List<Pair<String, List<Customer>>> customerEachCountryOrdersFirstCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersFirst = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(
                        c -> c.getOrders().size() == g.getValue().stream().findFirst().get().getOrders().size()).toList())).toList();
        return customerOrdersFirst;
    }

    public Map<String, List<Customer>> customerEachCountryOrdersFirstCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersFirst = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g->g.getValue().stream().filter(
                        c -> c.getOrders().size() == g.getValue().stream().findFirst().get().getOrders().size()).toList()));
        return customerOrdersFirst;
    }

    public int numbersFirstOrDefaultOrDefault() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).findFirst().orElse(-1);
        return firstNum;
    }

    public int bigNumbersFirstOrDefault() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).filter(n -> n > 5).findFirst().orElse(0);
        return firstNum;
    }

    public int oddNumbersFirstOrDefault() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).filter(n -> n % 2 == 1).findFirst().orElse(-2);
        return firstNum;
    }

    public double doubleFirstOrDefault() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).findFirst().orElse(3.14);
        return firstNum;
    }

    public double bigDoubleNumbersFirstOrDefault() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).filter(d -> d > 5).findFirst().orElse(-2.3);
        return firstNum;
    }

    public String stringFirstOrDefaultStartWithO() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startWithO = Arrays.stream(strings).filter(s -> s.charAt(0) == 'o').findFirst().orElse("none");
        return startWithO;
    }

    public String stringFirstOrDefaultStartWith(char c) {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startWithChar = Arrays.stream(strings).filter(s -> s.charAt(0) == c).findFirst().orElse("none");
        return startWithChar;
    }

    public String stringFirstOrDefaultLong() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        String startsWithO = Arrays.stream(strings).filter(s -> s.length() > 4).findFirst().orElse("empty");
        return startsWithO;
    }

    public int projectedBigNumbersFirstOrDefault() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int firstNum = Arrays.stream(numbers).map(i -> (int) Math.pow(2, i)).filter(n -> n > 100).findFirst().orElse(100);
        return firstNum;
    }

    public double projectedDoubleFirstOrDefault() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double firstNum = Arrays.stream(numbers).map(i -> Math.log(i)).filter(n -> n > 3).findFirst().orElse(3.01);
        return firstNum;
    }

    public int projectedStringFirstOrDefault() {
        String[] strings = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        int firstStringLength = Arrays.stream(strings).mapToInt(s -> s.length()).filter(i -> i % 2 == 0).findFirst().orElse(-1);
        return firstStringLength;
    }

    public Product productFirstOrDefault() {
        List<Product> products = getProductList();
        Product product12 = products.stream().filter(p -> p.getProductId() == 12).findFirst().orElse(Product.DEFAULT);
        return product12;
    }

    public List<Pair<String, Product>> productEachCategoryFirstOrDefault() {
        List<Product> products = getProductList();
        List<Pair<String, Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().findFirst().orElse(Product.DEFAULT))).toList();
        return categories;
    }

    public Map<String, Product> productEachCategoryFirstOrDefaultDict() {
        List<Product> products = getProductList();
        Map<String, Product> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().findFirst().orElse(Product.DEFAULT)));
        return categories;
    }

    public List<Pair<String, Order>> customerEachRegionFirstOrDefaultOrder() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Order>> ordersFirstOrDefault = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().flatMap(c -> c.getOrders().stream()).findFirst().orElse(Order.DEFAULT))).toList();
        return ordersFirstOrDefault;
    }

    public Map<String, Order> customerEachRegionFirstOrDefaultOrderDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Order> ordersFirstOrDefault = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().flatMap(c -> c.getOrders().stream()).findFirst().orElse(Order.DEFAULT)));
        return ordersFirstOrDefault;
    }

    public List<Pair<String, List<Product>>> productEachCategoryFirstOrDefaultUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(p ->
                        p.getUnitPrice() == g.getValue().stream().findFirst().orElse(Product.DEFAULT).getUnitPrice()).toList())).toList();
        return categories;
    }

    public Map<String, List<Product>> productEachCategoryFirstOrDefaultUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().filter(p ->
                                p.getUnitPrice() == g.getValue().stream().findFirst().orElse(Product.DEFAULT).getUnitPrice()).toList()));
        return categories;
    }

    public List<Pair<String, List<Customer>>> customerEachCountryOrdersFirstOrDefaultCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersFirstOrDefault = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(
                        c -> c.getOrders().size() == g.getValue().stream().findFirst().orElse(Customer.DEFAULT).getOrders().size()).toList())).toList();
        return customerOrdersFirstOrDefault;
    }

    public Map<String, List<Customer>> customerEachCountryOrdersFirstOrDefaultCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersFirstOrDefault = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g->g.getValue().stream().filter(
                        c -> c.getOrders().size() == g.getValue().stream().findFirst().orElse(Customer.DEFAULT).getOrders().size()).toList()));
        return customerOrdersFirstOrDefault;
    }

    public int NumbersElementAt() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int numAt = Arrays.stream(numbers).skip(1).findFirst().getAsInt();
        return numAt;
    }

    public int OddNumbersElementAt() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int numAt = Arrays.stream(numbers).filter(n -> n % 2 == 1).skip(3).findFirst().getAsInt();
        return numAt;
    }

    public double BigDoubleNumbersElementAt() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double numAt = Arrays.stream(numbers).filter(d -> d>5).skip(1).findFirst().getAsDouble();
        return numAt;
    }

    public String StringElementAtStartWithO() {
        String[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        String startWithO = Arrays.stream(strings).filter(s ->s.charAt(0)=='o').skip(0).findFirst().get();
        return startWithO;
    }

    public String StringElementAtStartWith(char c) {
        String[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        String startWithO = Arrays.stream(strings).filter(s ->s.charAt(0)==c).skip(2).findFirst().get();
        return startWithO;
    }

    public double ProjectedDoubleElementAt() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double numAt = Arrays.stream(numbers).map(i -> Math.log(i)).filter(n -> n > 3).skip(2).findFirst().getAsDouble();
        return numAt;
    }

    public int ProjectedStringElementAt() {
        String[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        int firstStringLength = Arrays.stream(strings).mapToInt(s -> s.length())
            .filter(i -> i % 2 == 0).skip(2).findFirst().getAsInt();
        return firstStringLength;
    }

    public Product ExpensiveProductElementAt() {
        List<Product> products = getProductList();
        Product expensiveProduct = products.stream().filter(p->p.getUnitPrice()>100)
                .skip(2).findFirst().get();
        return expensiveProduct;
    }

    public List<Pair<String, Product>> ProductEachCategoryElementAt() {
        List<Product> products = getProductList();
        List<Pair<String, Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
            .entrySet().stream().map(g -> Pair.with(g.getKey(),
                g.getValue().stream().skip(1).findFirst().get())).toList();
        return categories;
    }

    public Map<String, Product> ProductEachCategoryElementAtDict() {
        List<Product> products = getProductList();
        Map<String, Product> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g->g.getValue().stream().skip(1).findFirst().get()));
        return categories;
    }
}

File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/FirstMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class FirstMethods {
    //original:5 target:50 form ElementOperators
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public int NumbersFirst() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.First();
        return firstNum;
    }

    public int BigNumbersFirst() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.First(n => n > 5);
        return firstNum;
    }

    public int OddNumbersFirst() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.First(n => n % 2 == 1);
        return firstNum;
    }

    public double DoubleFirst() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.First();
        return firstNum;
    }

    public double BigDoubleNumbersFirst() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.First(d => d > 5);
        return firstNum;
    }

    public string StringFirstStartWithO() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithO = strings.First(s => s[0] == 'o');
        return startWithO;
    }

    public string StringFirstStartWith(char c) {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithChar = strings.First(s => s[0] == c);
        return startWithChar;
    }

    public string StringFirstLong() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startsWithO = strings.First(s => s.Length > 4);
        return startsWithO;
    }

    public int ProjectedBigNumbersFirst() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.Select(i => (int)Math.Pow(2, i)).First(n => n > 100);
        return firstNum;
    }

    public double ProjectedDoubleFirst() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.Select(i => Math.Log(i)).First(n => n > 3);
        return firstNum;
    }

    public int ProjectedStringFirst() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        int firstStringLength = strings.Select(s => s.Length).First(i => i % 2 == 0);
        return firstStringLength;
    }

    public Product ProductFirst() {
        List<Product> products = GetProductList();
        Product product12 = products.First(p => p.ProductId == 12);
        return product12;
    }

    public List<Tuple<string, Product>> ProductEachCategoryFirst() {
        List<Product> products = GetProductList();
        List<Tuple<string, Product>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.First())).ToList();
        return categories;
    }

    public Dictionary<string, Product> ProductEachCategoryFirstDict() {
        List<Product> products = GetProductList();
        Dictionary<string, Product> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.First());
        return categories;
    }

    public List<Tuple<string, Order>> CustomerEachRegionFirstOrder() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, Order>> ordersMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c => c.Orders).First())).ToList();
        return ordersMax;
    }

    public Dictionary<string, Order> CustomerEachRegionFirstOrderDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, Order> ordersMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.SelectMany(c => c.Orders).First());
        return ordersMax;
    }
    
    public List<Tuple<string, List<Product>>> ProductEachCategoryFirstUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.First().UnitPrice)
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ProductEachCategoryFirstUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.First().UnitPrice)
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Customer>>> CustomerEachCountryOrdersFirstCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersFirst = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Count ==g.First().Orders.Count).ToList())).ToList();
        return customerOrdersFirst;
    }

    public Dictionary<string, List<Customer>> CustomerEachCountryOrdersFirstCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersFirst = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Count ==g.First().Orders.Count).ToList());
        return customerOrdersFirst;
    }
    
    public int NumbersFirstOrDefault() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.FirstOrDefault(-1);
        return firstNum;
    }

    public int BigNumbersFirstOrDefault() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.FirstOrDefault(n => n > 5,0);
        return firstNum;
    }

    public int OddNumbersFirstOrDefault() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.FirstOrDefault(n => n % 2 == 1,-2);
        return firstNum;
    }

    public double DoubleFirstOrDefault() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.FirstOrDefault(3.14);
        return firstNum;
    }

    public double BigDoubleNumbersFirstOrDefault() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.FirstOrDefault(d => d > 5,-2.3);
        return firstNum;
    }

    public string StringFirstOrDefaultStartWithO() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithO = strings.FirstOrDefault(s => s[0] == 'o',"none");
        return startWithO;
    }

    public string StringFirstOrDefaultStartWith(char c) {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithChar = strings.FirstOrDefault(s => s[0] == c,"none");
        return startWithChar;
    }

    public string StringFirstOrDefaultLong() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startsWithO = strings.FirstOrDefault(s => s.Length > 4,"empty");
        return startsWithO;
    }

    public int ProjectedBigNumbersFirstOrDefault() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int firstNum = numbers.Select(i => (int)Math.Pow(2, i))
            .FirstOrDefault(n => n > 100,100);
        return firstNum;
    }

    public double ProjectedDoubleFirstOrDefault() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double firstNum = numbers.Select(i => Math.Log2(i))
            .FirstOrDefault(n => n > 3,3.01);
        return firstNum;
    }

    public int ProjectedStringFirstOrDefault() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        int firstStringLength = strings.Select(s => s.Length)
            .FirstOrDefault(i => i % 2 == 0,-1);
        return firstStringLength;
    }

    public Product ProductFirstOrDefault() {
        List<Product> products = GetProductList();
        Product product12 = products.FirstOrDefault(p => p.ProductId == 12,Product.Default);
        return product12;
    }

    public List<Tuple<string, Product>> ProductEachCategoryFirstOrDefault() {
        List<Product> products = GetProductList();
        List<Tuple<string, Product>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, 
                g.FirstOrDefault(Product.Default))).ToList();
        return categories;
    }

    public Dictionary<string, Product> ProductEachCategoryFirstOrDefaultDict() {
        List<Product> products = GetProductList();
        Dictionary<string, Product> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g =>
                g.FirstOrDefault(Product.Default));
        return categories;
    }

    public List<Tuple<string, Order>> CustomerEachRegionFirstOrDefaultOrder() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, Order>> ordersMax = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(p => p.Orders).FirstOrDefault(Order.Default))).ToList();
        return ordersMax;
    }

    public Dictionary<string, Order> CustomerEachRegionFirstOrDefaultOrderDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, Order> ordersMax = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.SelectMany(p => p.Orders).FirstOrDefault(Order.Default));
        return ordersMax;
    }
    
    public List<Tuple<string, List<Product>>> ProductEachCategoryFirstOrDefaultUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.FirstOrDefault(Product.Default).UnitPrice)
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ProductEachCategoryFirstOrDefaultUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.FirstOrDefault(Product.Default).UnitPrice)
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Customer>>> CustomerEachCountryOrdersFirstOrDefaultCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersFirst = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Count ==g.FirstOrDefault(Customer.Default).Orders.Count).ToList())).ToList();
        return customerOrdersFirst;
    }

    public Dictionary<string, List<Customer>> CustomerEachCountryOrdersFirstOrDefaultCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersFirst = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Count ==g.FirstOrDefault(Customer.Default).Orders.Count).ToList());
        return customerOrdersFirst;
    }
    
    public int NumbersElementAt() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int numAt = numbers.ElementAt(1);
        return numAt;
    }

    public int OddNumbersElementAt() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int numAt = numbers.Where(n => n % 2 == 1).ElementAt(3);
        return numAt;
    }

    public double BigDoubleNumbersElementAt() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double numAt = numbers.Where(d => d > 5).ElementAt(1);
        return numAt;
    }

    public string StringElementAtStartWithO() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithO = strings.Where(s => s[0] == 'o').ElementAt(0);
        return startWithO;
    }

    public string StringElementAtStartWith(char c) {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        string startWithChar = strings.Where(s => s[0] == c).ElementAt(2);
        return startWithChar;
    }

    public double ProjectedDoubleElementAt() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double numAt = numbers.Select(i => Math.Log(i)).Where(n => n > 3).ElementAt(2);
        return numAt;
    }
    
    public int ProjectedStringElementAt() {
        string[] strings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        int firstStringLength = strings.Select(s => s.Length)
            .Where(i => i % 2 == 0).ElementAt(2);
        return firstStringLength;
    }
    
    public Product ExpensiveProductElementAt() {
        List<Product> products = GetProductList();
        Product expensiveProduct = products.Where(p=>p.UnitPrice>100).ElementAt(2);
        return expensiveProduct;
    }
    
    public List<Tuple<string, Product>> ProductEachCategoryElementAt() {
        List<Product> products = GetProductList();
        List<Tuple<string, Product>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, 
                g.ElementAt(1))).ToList();
        return categories;
    }

    public Dictionary<string, Product> ProductEachCategoryElementAtDict() {
        List<Product> products = GetProductList();
        Dictionary<string, Product> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g =>
                g.ElementAt(1));
        return categories;
    }
}